src/transport/ws.rs

```diff
@@ -0,0 +1,335 @@
+        WsTransport {
+            stream: Arc::new(Mutex::new(WsStreamWrapper::Http((WebSocket::from_raw_socket(upgraded, Role::Server, None), addr)))),
+        }
+    }
+
+    pub fn new_tls(stream: WebSocket<AutoStream>, addr: Option<SocketAddr>) -> Self {
+        WsTransport {
+            stream: Arc::new(Mutex::new(WsStreamWrapper::Tls((stream, addr)))),
+        }
+    }
+}
+
+impl Read for WsTransport {
+    fn read(&mut self, mut buf: &mut [u8]) -> io::Result<usize> {
+        match self.stream.try_lock() {
+            Ok(mut stream) => stream.read(&mut buf),
+            Err(_) => Err(io::Error::new(io::ErrorKind::WouldBlock, "".to_string())),
+        }
+    }
+}
+
+impl AsyncRead for WsTransport {}
+
+impl Write for WsTransport {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        match self.stream.try_lock() {
+            Ok(mut stream) => stream.write(&buf),
+            Err(_) => Err(io::Error::new(io::ErrorKind::WouldBlock, "".to_string())),
+        }
+    }
+    fn flush(&mut self) -> io::Result<()> {
+        match self.stream.try_lock() {
+            Ok(mut stream) => stream.flush(),
+            Err(_) => Err(io::Error::new(io::ErrorKind::WouldBlock, "".to_string())),
+        }
+    }
+}
+
+impl AsyncWrite for WsTransport {
+    fn shutdown(&mut self) -> Result<Async<()>, std::io::Error> {
+        match self.stream.try_lock() {
+            Ok(mut stream) => stream.async_shutdown(),
+            Err(_) => Err(io::Error::new(io::ErrorKind::WouldBlock, "".to_string())),
+        }
+    }
+}
+
+impl Transport for WsTransport {
+    fn connect(url: &Url) -> JetFuture<Self>
+        where
+            Self: Sized,
+    {
+        let addr = url_to_socket_arr(url);
+        let owned_url = url.clone();
+        Box::new(futures::lazy(move || {
+            tungstenite::connect(owned_url).map(|(stream, _)| {
+                WsTransport::new_tls(stream, Some(addr))
+            }).map_err(|e| io::Error::new(io::ErrorKind::Other, e.to_string()))
+        })) as JetFuture<Self>
+    }
+
+    fn message_sink(&self) -> JetSinkType<Vec<u8>> {
+        Box::new(WsJetSink::new(self.stream.clone()))
+    }
+
+    fn message_stream(&self) -> JetStreamType<Vec<u8>> {
+        Box::new(WsJetStream::new(self.stream.clone()))
+    }
+}
+
+struct WsJetStream {
+    stream: Arc<Mutex<WsStreamWrapper>>,
+    nb_bytes_read: u64,
+    packet_interceptor: Option<Box<dyn PacketInterceptor>>,
+}
+
+impl WsJetStream {
+    fn new(stream: Arc<Mutex<WsStreamWrapper>>) -> Self {
+        WsJetStream {
+            stream,
+            nb_bytes_read: 0,
+            packet_interceptor: None,
+        }
+    }
+}
+
+impl Stream for WsJetStream {
+    type Item = Vec<u8>;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Result<Async<Option<<Self as Stream>::Item>>, <Self as Stream>::Error> {
+        if let Ok(ref mut stream) = self.stream.try_lock() {
+            let mut buffer = [0u8; 65535];
```

Comment:
is this the best way to handle receiving bytes? it creates a 65535-byte buffer, sets it to zero, then a bunch of bytes are read into it, and the result is converted to a vector.

Reply:
I would recommend at least using a Vec as static arrays are allocated on the stack.

Reply:
This was copy pasted from the src/transport/tcp.rs file, if this is to be changed, it should be changed there as well.

Reply:
Oops :) I'll ask François about it then

Reply:
The match on stream.poll_read(&mut buffer) could be done in a loop where you would read small amount until nothing is left to read / or until max tcp frame is reached (e.g. 65535). The fact that each time poll is called a 65KB array / vec is allocated and then truncated to a potentially way-shorter one is also something that would be fixed by a loop.